include/tvm/node/container.h

```diff
@@ -1449,31 +1449,41 @@ namespace runtime {
 // Additional overloads for PackedFunc checking.
 template <typename T>
 struct ObjectTypeChecker<Array<T>> {
-  static bool Check(const Object* ptr) {
-    if (ptr == nullptr) return true;
-    if (!ptr->IsInstance<ArrayNode>()) return false;
+  static Optional<String> Mismatch(const Object* ptr) {
+    if (ptr == nullptr) return Optional<String>();
+    if (!ptr->IsInstance<ArrayNode>()) return Optional<String>(ptr->GetTypeKey());
     const ArrayNode* n = static_cast<const ArrayNode*>(ptr);
-    for (const ObjectRef& p : *n) {
-      if (!ObjectTypeChecker<T>::Check(p.get())) {
-        return false;
+    for (size_t i = 0; i < n->size(); i++) {
+      const ObjectRef& p = (*n)[i];
+      auto check_subtype = ObjectTypeChecker<T>::Mismatch(p.get());
```

Comment:
Be more explicit about the return type. My (personal) rule of thumb is that if it is not c++ iterator, and the type does not appear on the same line, then I would force myself to write it out

```suggestion
      Optional<String> check_subtype = ObjectTypeChecker<T>::Mismatch(p.get());
```